File: datagen-stream-fighter/src/main/scala/com/github/simplesteph/udemy/scala/datagen/Dataset.scala
package com.github.simplesteph.udemy.scala.datagen

import java.time.Instant

object Dataset {

  sealed abstract class Character(val name: String, val country: String)

  case object KEN extends Character("Ken", "US")
  case object RYU extends Character("Ryu", "Japan")
  case object GEKI extends Character("Geki", "Japan")
  case object CHUNLI extends Character("ChunLi", "China")
  case object AKUMA extends Character("Akuma", "Japan")
  case object SAKURA extends Character("Sakura", "Japan")
  case object DHALSIM extends Character("Dhalsim", "India")
  case object BLAIR extends Character("Blair", "UK")
  case object BLANKA extends Character("BLANKA", "Brazil")

  val StreetFighterCast: Vector[Character] = Vector(RYU, KEN, CHUNLI, GEKI, AKUMA, SAKURA, DHALSIM, BLAIR)

  sealed trait Game

  case object StreetFighter extends Game
  case object Takken extends Game
  case object KingOfFighters extends Game
  case object SoulCalibur extends Game
  case object SamuraiShodown extends Game

  val GameCollection: Vector[Game] = Vector(StreetFighter, Takken, KingOfFighters, SoulCalibur, SamuraiShodown)

  case class User(login: String, firstName: String, lastName: Option[String] = None, email: Option[String] = None)



  /* helpers */

  object User {

    def apply(login: String, firstName: String, lastName: String, email: String): User =
      new User(login, firstName, Some(lastName), Some(email))

    def apply(login: String, firstName: String, lastName: String): User =
      new User(login, firstName, Some(lastName))
  }

  case class Purchase(user: String, game: Game, twoPlayer: Boolean)

  object ExactlyOnceExercise {

    case class Challenger(login: String, character: String)

    object Challenger {
      def apply(login: String, character: Character): Challenger =
        new Challenger(login, character.name)
    }
    case class Hit(key: String, challenger: Challenger, damage: Int, time: Instant)

  }
}

File: datagen-stream-fighter/src/main/java/com/github/simplesteph/udemy/java/datagen/Dataset.java
package com.github.simplesteph.udemy.java.datagen;


import org.apache.avro.Schema;
import org.apache.avro.SchemaBuilder;

public class Dataset {

    public static Schema UserSchema = SchemaBuilder.record("User")
            .namespace("com.github.simplesteph")
            .fields()
            .name("login").type().stringType().noDefault()
            .name("first-name").type().stringType().noDefault()
            .name("last-name").type().nullable().stringType().noDefault()
            .name("email").type().nullable().stringType().noDefault()
            .endRecord();

    public static Schema UserKeySchema = SchemaBuilder.record("UserKey")
            .namespace("com.github.simplesteph")
            .fields()
            .name("login").type().stringType().noDefault()
            .endRecord();

    public static Schema PurchaseSchema = SchemaBuilder.record("Purchase")
            .namespace("com.github.simplesteph")
            .fields()
            .name("user").type().stringType().noDefault()
            .name("game").type().enumeration("Game").namespace("com.github.simplesteph")
            .symbols("", "", "").noDefault()
            .name("two-player").type().booleanType().booleanDefault(false)
            .endRecord();

    public static Schema PurchaseKeySchema = SchemaBuilder.record("User")
            .namespace("com.github.simplesteph")
            .fields()
            .name("id").type().stringType().noDefault()
            .name("client").type((UserSchema)).noDefault()
            .endRecord();

}
